clientpool/channel.go

```diff
@@ -17,6 +17,10 @@ type channelPool struct {
 var _ Pool = (*channelPool)(nil)
 
 // NewChannelPool creates a new client pool implemented via channel.
+//
+// Note that this function could return both non-nil Pool and error,
+// When we failed to create all asked initialClients.
+// In such case the returned Pool would have the clients we already established.
```

Comment:
This seems like it's making the API harder to use correctly -- this is likely to lead to un-Close-d pools when we return an error.  You will have to say
```go
pool, err := NewChannelPool(...)
if pool != nil {
  defer pool.Close()
}
if err != nil {
  ...
}
```

which is not very natural.

If we're going to return a Pool, I think we need to _not_ return an error.

Reply:
there are a few instances in the stdlib that the function can return both non-nil result and error, although the only one non-nil pattern is far more common.

but I like your suggestion of changing the config to a required int, so I'll do that instead.

Reply:
but on second thought, that means we are losing the info about how the M-th connection was failed to be established (unless we make a breaking change to also pass in a logger to log that, or something like that).

`clientpool` is considered a low-level api that most services will not use it directly (internal code search shows only 2 repo uses it, see my PR description/commit message), how strongly do you feel that we should just swallow the error and don't return an error?

Reply:
doesn't the logger satisfy that need of wanting to control where and how errors are handled?

Reply:
we don't have logger in this API now. adding it will be a breaking change.

in my experience that error usually can provide info to shed light on whether the error is on dns or on the remote service.

Reply:
I meant `InitialConnectionsFallbackLogger`?  Since you're going to be _trying_ to stand up the connections, some will just fail, and they'll be logged there.

The workflow you're describing is primarily human driven, so the `error` isn't necessary as long as it can be logged.

Reply:
that's part of the thriftbp api, not part of the clientpool api (clientpool is the more low-level api called by thriftbp)